pkg/grpc_middleware/dialer/builder.go

```diff
@@ -0,0 +1,319 @@
+package dialer
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"log"
+	"net/url"
+	"strconv"
+	"time"
+
+	"github.com/caring/go-packages/v2/pkg/errors"
+	grpc_middleware "github.com/grpc-ecosystem/go-grpc-middleware"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/backoff"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/keepalive"
+)
+
+const (
+	// MinConnectTimeout before attempting reconnect.
+	MinConnectTimeout = 20 * time.Second
+)
+
+type Builder struct {
+	options         []grpc.DialOption
+	enabledBlocking bool
+	connectParams   grpc.ConnectParams
+	keepAliveParams keepalive.ClientParameters
+	credentials     credentials.PerRPCCredentials
+	tlsConfig       *tls.Config
+	uinterceptors   []grpc.UnaryClientInterceptor
+	sinterceptors   []grpc.StreamClientInterceptor
+	dns             *string
+	port            *string
+	withTLS         *bool
+}
+
+// WithOptions allows possing in multiple grpc dial options
+// DialOption configures how we set up the connection.
+func (b *Builder) WithOptions(opts ...grpc.DialOption) {
+	b.options = opts
+}
+
+// AppendOptions appends the given options to the current set of options.
+func (b *Builder) AppendOptions(opts ...grpc.DialOption) {
+	b.options = append(b.options, opts...)
+}
+
+// GetOptions returns the current set of options
+func (b *Builder) GetOptions() []grpc.DialOption {
+	return b.options
+}
+
+// WithBlock set to true applies a DialOption which makes caller of Dial blocks until the
+// underlying connection is up. Without this, Dial returns immediately and
+// connecting the server happens in background.
+func (b *Builder) WithBlock(blocking bool) {
```

Comment:
I think these should be eliminated in favor of just using the grpc dial option directly.

Reply:
for less experienced engineers, I think the flexibility of DialOptions can be complicated and difficult to understand... one of the original intentions of this was to simplify that into some "obvious" configurations that you just called methods for... while still providing the flexibility of using DialOptions directly...

Reply:
That's a fair point. Though we have since learned that this option, while useful, isn't desired most of the time in production. Would keeping this and converting the non-specific dial option methods to public fields might be a good combo?